### PR TITLE
Return an error if asked to wait on completed future

### PR DESCRIPTION
**Pull request by riyafa**
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java [hunk 1]
```diff
@@ -322,10 +336,12 @@ public WaitResult handleWaitAny(List<FutureValue> futures) throws Throwable {
         return waitResult;
     }
 
+    private BError createWaitOnSameFutureError() {
+        return ErrorCreator.createError(StringUtils.fromString("multiple waits on the same future is not allowed"));
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java:340]: Shall we move this error to https://github.com/ballerina-platform/ballerina-lang/blob/master/bvm/ballerina-runtime/src/main/resources/MessagesBundle.properties.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/future/FutureNegativeTests.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
+/*
+ *  Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/future/FutureNegativeTests.java:2]:
```suggestion
 *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

**Comment by warunalakshitha** [comment]
Related compiler FE PR https://github.com/ballerina-platform/ballerina-lang/pull/30456

File: tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        panic r1;
+    }
+}
+
+public function testMultipleWaitTwiceOnDifferentFutures() {
+    future<int> f1 = start calculate(3, 4);
+    future<int|error> f4 = start calculate(9, 8);
+    future<int|error> f5 = start calculate(5, 4);
+    record { int|error r1; int|error r2; } result = wait {r1: f4, r2: f5};
+    record { int|error r1; int|error r2; } result1 = wait {r1: f4, r2: f1};
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal:43]: Shall we check the  r2 value, which should not be an error?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        panic r1;
+    }
+}
+
+public function testAlternateWaitTwiceOnTwoFutures() {
+    future<int|error> f4 = start calculate(9, 8);
+    future<int|error> f5 = start calculate(5, 4);
+    int|error result = wait f4|f5;
+    int|error r1 = wait f4|f5;
+    if(r1 is error) {
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal:55]: Refer example.
``` 
future<int|error> f4 = start calculate(9, 8);
future<int|error> f5 = start calculate(5, 4);
int|error result = wait f4;
```

case 01  result = wait f4|f5; -> error 
case 02  result = wait f5|f4; -> future
 
What would be the expected outcome for two cases?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java [hunk 5]
```diff
@@ -290,14 +299,19 @@ public WaitResult handleWaitAny(List<FutureValue> futures) throws Throwable {
                         ctx.unLock();
                         throw future.panic;
                     }
-
+                    if (future.hasWaited()) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java:302]: Should we move this logic to the future.getResult method? So we do not need to check to wait in multiple places.
What happens if somebody passes already waited for the future to interop method and call get a result.

ex.
    public Object getResult() {
        if(waited){
            return createWaitOnSameFutureError();
        }
        waited = true
       return this.result;
    }

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java [hunk 6]
```diff
@@ -256,14 +261,22 @@ public void handleWaitMultiple(Map<String, FutureValue> keyValues, MapValue targ
...
-                target.put(StringUtils.fromString(entry.getKey()), future.result);
+                doneCount++;
             } else {
                 this.setState(BLOCK_ON_AND_YIELD);
                 entry.getValue().strand.waitingContexts.add(ctx);
             }
             future.strand.unlock();
         }
+        if (doneCount == keyValues.size()) {
+            keyValues.forEach((s, futureValue) -> futureValue.setWaited(true));
```
Comment (manuranga) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java:278]: let's use a simple loop, we have to convention of using raw loops for performance/debug critical paths.

Comment (manuranga) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java:278]: why is the setWaited happening  after unlocking, want this cause a issue?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal (hunk 3 above)
Comment (riyafa) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal:48]: I have re written the tests to check this as well

File: tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal (hunk 4 above)
Comment (riyafa) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal]: In the above cases a result would be returned. They wouldn't error

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java (hunk 5 above)
Comment (riyafa) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java:307]: I think we discussed this on call

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java (hunk 6 above)
Comment (riyafa) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java:278]: Created an issue to handle potential concurrency issues: https://github.com/ballerina-platform/ballerina-lang/issues/30621

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Pull request merged**